presto-hive/src/main/java/com/facebook/presto/hive/HivePartitionManager.java

```diff
@@ -121,18 +122,23 @@ public HivePartitionResult getPartitions(SemiTransactionalHiveMetastore metastor
                 .map(column -> typeManager.getType(column.getTypeSignature()))
                 .collect(toList());
 
-        Iterable<HivePartition> partitionsIterable = partitionColumns.isEmpty() ? ImmutableList.of(new HivePartition(tableName)) : () -> getFilteredPartitionNames(metastore, tableName, partitionColumns, effectivePredicate).stream()
+        List<HivePartition> partitions = partitionColumns.isEmpty() ? ImmutableList.of(new HivePartition(tableName)) : getFilteredPartitionNames(metastore, tableName, partitionColumns, effectivePredicate).stream()
```

Comment:
Curious: This looks unrelated change? :) . In theory this can increase the memory usage on coordinator since now it eagerly constructs `HivePartition`. 

For example, this part of code guarantees we never have too many `HivePartition` object instantiate:  https://github.com/prestodb/presto/blob/903e19c46ea7d642672c0b5b25d520689e2d9f6e/presto-hive/src/main/java/com/facebook/presto/hive/HiveMetadata.java#L621-L627

Reply:
it's not unrelated.  I'm trying to get the number of partitions, and the final number isn't available until after `parseValuesAndFilterPartition` constructs the partitions. I can look more into why those filters can't be applied before creating the partition or also use the larger number that comes from "getFilteredPartitionNames", but that could overestimate.

Reply:
just looked at the code. it's not really possible to get the number of partitions before listing them all.

Reply:
I've been thinking about this more and i'm not sure this change is a good idea. Given that this config is specifically intended queries that have lots of partitions * buckets so they don't have to materialize them all at once.  It seems like now we're just moving the problem elsewhere for queries with lots of partitions. @wenleix @arhimondr  what do you think?

Reply:
From what I understand, returning an iterator only makes sense if it is used in the `getPartitionsSystemTable` (well, it kinda makes sense, because if there are too many partitions queries, the query will rather fail on the `metastore.getPartitionNamesByParts`).

In other cases partitions are being materialized to a list by the `getPartitionsAsList` anyway (yes, enforcing some limit).

If we move the materialization and the partitions limit check into the `HivePartitionManager#getPartitions` - it won't change anything for the second case.

For the first case i would suggest to add a second method to the `HivePartitionManager`. e.g.: `Iterable<HivePartition> getPartitionsIterator(...)` and use it in the `getPartitionsSystemTable`.

Reply:
@arhimondr can you look at the new commit?